src/bun.js/node/types.zig

```diff
@@ -1835,548 +1849,2876 @@ pub const Emitter = struct {
+                end = len;
+            } else if (start == end.?) {
+                end = firstNonSlashEnd;
+            }
+
+            return path[start ..end.?];
         }
 
-        // "a/b" => "a" or "//b" => "//"
-        if (end_index <= 1) {
-            if (path[0] == '/' and path[1] == '/') {
-                end_index += 1;
+        // Add a block to isolate `i`.
+        {
+            var i: usize = len - 1;
+            while (i >= start) : (i -= if (i > 0) 1 else break) {
+                const byte = path[i];
+                if (isSepImpl(T, byte)) {
+                    if (!matchedSlash) {
+                        start = i + 1;
+                        break;
+                    }
+                } else if (end == null) {
+                    matchedSlash = false;
+                    end = i + 1;
+                }
             }
         }
 
-        return path[0..end_index];
-    }
-
-    pub fn dirname(globalThis: *JSC.JSGlobalObject, isWindows: bool, args_ptr: [*]JSC.JSValue, args_len: u16) callconv(.C) JSC.JSValue {
-        if (comptime is_bindgen) return JSC.JSValue.jsUndefined();
-        if (args_len == 0) {
-            return JSC.toInvalidArguments("path is required", .{}, globalThis);
+        if (end == null) {
+            return "";
         }
-        var stack_fallback = std.heap.stackFallback(4096, JSC.getAllocator(globalThis));
-        const allocator = stack_fallback.get();
-
-        var arguments: []JSC.JSValue = args_ptr[0..args_len];
-        var path = arguments[0].toSlice(globalThis, allocator);
-        defer path.deinit();
 
-        const base_slice = path.slice();
+        return path[start ..end.?];
+    }
 
-        const out = if (isWindows)
-            @This().dirnameWindows(base_slice)
-        else
-            @This().dirnamePosix(base_slice);
+    pub inline fn basenamePosix(path: []const u8, suffix: ?[]const u8) []const u8 {
+        return basenamePosixImpl(u8, path, suffix);
+    }
 
-        return JSC.ZigString.init(out).withEncoding().toValueGC(globalThis);
+    pub inline fn basenameWindows(path: []const u8, suffix: ?[]const u8) []const u8 {
+        return basenameWindowsImpl(u8, path, suffix);
     }
 
-    pub fn extname(globalThis: *JSC.JSGlobalObject, _: bool, args_ptr: [*]JSC.JSValue, args_len: u16) callconv(.C) JSC.JSValue {
+    pub fn basename(globalThis: *JSC.JSGlobalObject, isWindows: bool, args_ptr: [*]JSC.JSValue, args_len: u16) callconv(.C) JSC.JSValue {
         if (comptime is_bindgen) return JSC.JSValue.jsUndefined();
-        if (args_len == 0) {
-            return JSC.toInvalidArguments("path is required", .{}, globalThis);
+        const suffix_ptr: ?JSC.JSValue = if (args_len > 1) args_ptr[1] else null;
+
+        if (suffix_ptr != null) {
+            // Supress exeption in zig. It does throw in JS land.
+            validateString(globalThis, suffix_ptr.?, "ext", .{}) catch {
+                return JSC.JSValue.jsUndefined();
```

Comment:
Compiler is happy with `.zero`. Less is more, I can dig it :)

Reply:
Scratch that. Turns out using `.zero` instead of `JSC.JSValue.jsUndefined();` causes grumpy segfaults 🙈

Reply:
interesting. there is also `.undefined` but im confused how zero crashes when there is also an exception present.

Reply:
It crashes because the method expected return value of JSValue and it didn't get it (being .zero). The trigger of the error in JS land happens after in order of operations.